app/services/migration/user_migrator.rb

```diff
@@ -0,0 +1,117 @@
+module Migration
+  class UserMigrator
+    attr_reader :ecf_user
+
+    def initialize(ecf_user)
+      @ecf_user = ecf_user
+    end
+
+    def find_initialize_or_merge_user
+      if user_does_not_exist?
+        return ::User.new(ecf_id: ecf_user.id)
+      end
+
+      validate_multiple_users_with_same_ecf_id
+
+      if both_ids_return_same_user? || only_ecf_id_return_user?
+        return ecf_user_id_user
+      end
+
+      if both_ids_return_different_users?
+        # Scenario 1 on JIRA
+        # - ecf_user1 => user1
+        # - ecf_user1 => user2
+        # - user2 => ecf_user2 (orphan or nil)
+        if gai_user_ecf_id_links_to_nothing? || gai_user_ecf_id_links_to_orphan_ecf_user?
+          # We merge `ecf_user_gai_id_user` into `ecf_user_id_user`
+          UserMerger.new(from_user: ecf_user_gai_id_user, to_user: ecf_user_id_user).merge!
+          return ecf_user_id_user
+        end
+
+        # Scenario 3 and Scenario 2 (when running on ecf_user with 8910)
+        # - ecf_user1 => user1
+        # - ecf_user1 => user2
+        # - user2 => ecf_user2
+        raise_error("ecf_user.id and ecf_user.get_an_identity_id both return NPQ User records, and the NPQ User link to non-orphan ECF users")
+      end
+
+      if only_gai_id_return_user?
+        if gai_user_ecf_id_links_to_nothing? || gai_user_ecf_id_links_to_orphan_ecf_user?
+          # We set `user.ecf_id` to `ecf_user.id`, return user
+          ecf_user_gai_id_user.ecf_id = ecf_user.id
+          return ecf_user_gai_id_user
+        end
+
+        # Scenario 2 on JIRA (when running on ecf_user with 123)
+        raise_error("User found with ecf_user.get_an_identity_id, but its user.ecf_id linked to another ecf_user that is not an orphan")
+      end
+    end
+
+  private
+
+    def ecf_user_id_user
```

Comment:
I think the method naming for retrieving NPQ users from ECF details are a bit confusing; I expected this one to return an ECF user at first glance. What about following ActiveRecord with something like `user_by_ecf_user_id` and `user_by_ecf_user_get_an_identity_id`?

Reply:
Yes, i started off with that naming but I didn't like it ending with `_id`. I'll rename them.